### ISSUE TITLE
add support for gridtables

### ISSUE THREAD

**Issue by tripodsan** [adobe/helix-markdown-support#122]
Research if https://github.com/zestedesavoir/zmarkdown/tree/master/packages/remark-grid-tables can be used for grid tables in helix

also see: https://pandoc.org/MANUAL.html#extension-grid_tables

GridTables look like this:

```
+-------------------+------+
| Table Headings    | Here |
+--------+----------+------+
| Sub    | Headings | Too  |
+========+=================+
| cell   | column spanning |
| spans  +---------:+------+
| rows   |   normal | cell |
+---v----+:---------------:+
|        | cells can be    |
|        | *formatted*     |
|        | **paragraphs**  |
|        | ```             |
| multi  | and contain     |
| line   | blocks          |
| cells  | ```             |
+========+=========<+======+
| footer |    cells |      |
+--------+----------+------+
```

- the top of a cell must be indicated by `+-` followed by some `-` or `+` and finished by `-+`.
- if the table contains a footer but no header, the top row should use `=` as grid line.
- col spans are indicated by missing column (`|`) delimiters
- row spans are indicated by missing row (`-`) delimiters
- cells can be left, center, right, or justify aligned; indicated by the placement of `:` or `><`
- cells can be top, middle, or bottom v-aligned; indicated by the placement of arrows (`v` `^` `x`)
- the header and footer sections are delimited by section delimiters (`=`).
- if no section delimiters are present, all cells are placed in the table body.
- if only 1 section delimiter is present, it delimits header from body.
- the content in cells can be a full Markdown document again. note, that the cell boundaries (`|`)
  need to exactly match with the column markers (`+`) in the row delimiters, if the cell content
  contains `|`, otherwise the correct layout of the table can't be guaranteed.

Layout
======

The table layout tries to keep the table within a certain width (default 120). For example,
if the table has 3 columns, each column will be max 40 characters wide. If all text in a column
is smaller, it will shrink the columns. However, cells have a minimum width (default 10) when
text needs to be broken. If the cell contents need more space, e.g. with a nested table or
code block, it will grow accordingly.

Align
=====

Horizontal align is indicated by placing markers at the grid line above the cell:

```
Justify     Center     Left       Right
+>-----<+  +:-----:+  +:------+  +------:+
| A b C |  |  ABC  |  | ABC   |  |   ABC |
+-------+  +-------+  +-------+  +-------+
```

Vertical align is indicated by placing markers at the center of the grid line above the cell:

```
Top        Middle     Bottom
+---^---+  +---x---+  +---v---+
| Larum |  |       |  |       |
| Ipsum |  | Larum |  |       |
|       |  | Ipsum |  | Larum |
|       |  |       |  | Ipsum |
+-------+  +-------+  +-------+
```

## Syntax

```ebfn

gridTable := tbd;

```

Note: the top of a cell must be indicated by `+-` followed by some `-` or `+` and finished by `-+`.
So, this is not a correct table:

```
+--+
|a |
+--+
```

But, this is a correct table:

```
+---+
| a |
+---+
```

## MDAST Syntax tree

The following interfaces are added to **[mdast][]** by this utility.

#### `GridTable`

```idl
interface GridTable <: Parent {
  type: "gridTable"
  children: [GridTableHeader|GridTableBody|GridTableFooter]
}
```

#### `GridTableHeader`

```idl
interface GridTableHeader <: Parent {
  type: "gtHead"
  children: [GridTableRow]
}
```

#### `GridTableBody`

```idl
interface GridTableBody <: Parent {
  type: "gtBody"
  children: [GridTableRow]
}
```

#### `GridTableFoot`

```idl
interface GridTableFooter <: Parent {
  type: "gtFoot"
  children: [GridTableRow]
}
```

#### `GridTableRow`

```idl
interface GridTableRow <: Parent {
  type: "gtRow"
  children: [GridTableCell]
}
```

#### `GridTableCell`

```idl
interface GridTableCell <: Parent {
  type: "gtCell"
  colSpan: number >= 1
  rowSpan: number >= 1
  align: alignType
  valign: valignType
  children: [MdastContent]
}
```

**GridTableCell** ([**Parent**][dfn-parent]) represents a header cell in a
[**GridTable**][dfn-table], if its parent is a [*gridTableHead*][term-head], or a data
cell otherwise.

**GridTableCell** can be used where [**gridTableRow**][dfn-row-content] content is expected.
Its content model is [**mdast**][dfn-phrasing-content] content, allowing full mdast documents.

#### `alignType`

```idl
enum alignType {
  "left" | "right" | "center" | "justify" | null
}
```

#### `valignType`

```idl
enum alignType {
  "top" | "bottom" | "middle" | null
}
```

**Comment by tripodsan** [comment]
progress:
<img width="608" alt="image" src="https://user-images.githubusercontent.com/917628/182316011-13826617-ff50-45c9-9899-5f42f4306cfa.png">

**Comment by tripodsan** [comment]
progress:

<img width="376" alt="image" src="https://user-images.githubusercontent.com/917628/182398817-4b14ff1e-f42d-4fc7-8f68-6bdba636937d.png">

**Comment by tripodsan** [comment]
parsed grid. probably needs 2 phases...

```
root[2]
├─0 heading[1]
│   │ depth: 2
│   └─0 text "Large Grid Table"
└─1 gridTable[1]
    └─0 gtHeader[3]
        ├─0 gtRow[3]
        │   ├─0 gtCell[1]
        │   │   └─0 text " A1                                                                                         "
        │   ├─1 gtCell[1]
        │   │   └─0 text " B1                                                           "
        │   └─2 gtCell[1]
        │       └─0 text " ## Lorem ipsum dolor sit amet, consectetur adipiscing elit. "
        ├─1 gtRow[3]
        │   ├─0 gtCell[15]
        │   │   ├─0  text " ## My Heading 1                                                                            "
        │   │   ├─1  text "                                                                                            "
        │   │   ├─2  text " ![](https://hlx.blob.core.windows.net/external/19c0cf25413106c81920d75078ee2ef30a55d52e7)\\ "
        │   │   ├─3  text " Lorem ipsum dolor **sit amet**, consectetur adipiscing elit. Vivamus rhoncus elit nibh,    "
        │   │   ├─4  text " sed vestibulum metus tincidunt a. _Integer_ interdum tempus consectetur. Phasellus         "
        │   │   ├─5  text " tristique auctor tortor, tincidunt semper odio blandit eu. Proin et aliquet est. Curabitur "
        │   │   ├─6  text " ac augue ornare, iaculis sem luctus, feugiat tellus.\\                                      "
        │   │   ├─7  text " 0 1 2 3 4 5 6 7 8 9 0 a b c d e f 0 1 2 3 4 5 6 7 8 9 0 a b c d e f                        "
        │   │   ├─8  text "                                                                                            "
        │   │   ├─9  text "                                                                                            "
        │   │   ├─10 text "                                                                                            "
        │   │   ├─11 text "                                                                                            "
        │   │   ├─12 text "                                                                                            "
        │   │   ├─13 text "                                                                                            "
        │   │   └─14 text "                                                                                            "
        │   ├─1 gtCell[15]
        │   │   ├─0  text " ```js                                                        "
        │   │   ├─1  text " for (const row of this.rows) {                               "
        │   │   ├─2  text "   for (let i = 0; i < row.length; i += 1) {                  "
        │   │   ├─3  text "     let col = cols[i];                                       "
        │   │   ├─4  text "     if (!col) {                                              "
        │   │   ├─5  text "       col = {};                                              "
        │   │   ├─6  text "       cols.push(col);                                        "
        │   │   ├─7  text "     }                                                        "
        │   │   ├─8  text "     const cell = row[i];                                     "
        │   │   ├─9  text "     if (cell.value) {                                        "
        │   │   ├─10 text "       col.size = Math.max(col.size || 0, cell.value.length); "
        │   │   ├─11 text "     }                                                        "
        │   │   ├─12 text "   }                                                          "
        │   │   ├─13 text " }                                                            "
        │   │   └─14 text " ```                                                          "
        │   └─2 gtCell[15]
        │       ├─0  text " > My quote                                                  "
        │       ├─1  text " >                                                           "
        │       ├─2  text " > Lorem ipsum dolor sit amet, consectetur adipiscing elit.  "
        │       ├─3  text " > Vivamus rhoncus elit nibh, sed vestibulum metus tincidunt "
        │       ├─4  text " > a.                                                        "
        │       ├─5  text "                                                             "
        │       ├─6  text " -   item one                                                "
        │       ├─7  text " -   item two                                                "
        │       ├─8  text " -   Lorem ipsum dolor sit amet, consectetur adipiscing      "
        │       ├─9  text "     elit. Vivamus rhoncus elit nibh, sed vestibulum metus   "
        │       ├─10 text "     tincidunt a.                                            "
        │       ├─11 text "                                                             "
        │       ├─12 text "                                                             "
        │       ├─13 text "                                                             "
        │       └─14 text "                                                             "
        └─2 gtRow[3]
            ├─0 gtCell[1]
            │   └─0 text " a3                                                                                         "
            ├─1 gtCell[1]
            │   └─0 text " b3                                                           "
            └─2 gtCell[1]
                └─0 text " c3                                                          "
```

**Comment by tripodsan** [comment]
closed with #126
